Simplify/automate the compilation and use of a system image

ffevotte opened the pull request:
The following PR introduces features helping compile and use custom system images for LanguageServer and SymbolServer. The expected benefits are reduced start-up times for the language server. On my (old-ish) machine, start-up times are reduced from 20 seconds to less than 2. On a newer machine where I also tested this, times went from 15 seconds down to less than 1.

The idea is the following:
- `eglot-jl.jl` looks at the `EGLOT_JL_TEST` environment variable; when it is set, the server is configured to read from an IOBuffer that sends it an `exit` instruction as soon as it is able to process requests. This allows `eglot-jl.jl` to be used as a `precompile_execution_file` that exercises the (language) server before gracefully exiting.
- a `compile.jl` script automates the generation of a system image, named after the julia version (and having the relevant extension depending on the OS).

This is what the first commit in this PR introduces, alongside with instructions in `README.org` explaining how to compile the system image, and set `eglot-jl-julia-args` to use it.

The second commit goes a step further: it automates the search for a suitable system image at server startup, in order to use it if one is found.
- a first call to julia allows getting the version number. This involves something like 0.1s additional latency before the server is run, which I think is acceptable but is nevertheless quite a large amount of time for such an insignificant-looking task
- a suitable system image is looked for, based on the version number; if one is found, the corresponding `--sysimage` command-line switch is generated. If not, nothing happens; no system image is used, so users who have not generated any system image will not be impacted by this.
- the whole process handling system images can be deactivated using a new `eglot-jl-enable-sysimage` customizable option. This allows users to opt out of the system, even if they have previously generated a system image (for example in case it would become stale for some reason). It also completely removes the julia version overhead.

Additionally, a new autoloaded command allows running the system image generation script from within emacs. The same command could be used to re-generate a system image if the previous one caused problems for any reason.

This is documented in `README.org`. I've tested this with Julia 1.3 and 1.4, on Windows and Linux, without encountering any problem (I don't have access to a Mac).

I know you were reluctant to handle system images in `eglot-jl`, and were (rightfully) especially concerned about possible ways that a system image would become obsolete. I tried making this implementation as robust as possible, but I might very well have overlooked something. So please do not hesitate to tell me if something bothers you with this proposal.

ffevotte committed:
Simplify compilation and use of system images

ffevotte committed:
Automatic detection of system images

The compilation process is also streamlined, allowing it to be run directly from
Emacs with `M-x eglot-jl-compile-sysimage`.

non-Jedi commented:
Could you rebase on master, and I'll try to review in the next few days?

ffevotte commented:
OK, the PR is now rebased on the current master, and everything seems to work on my system.

There were several conflicts though; I'll try and test more in the coming days, and report back if anything goes wrong.

non-Jedi commented on eglot-jl.jl:
```diff
@@ -30,8 +30,21 @@ project_path = something(Base.current_project(src_path), Base.load_path_expand(L
+               "id"      => 1,
+               "method"  => "exit",
+               "params"  => Dict()) |> JSON.json
+    input = IOBuffer("Content-Length: $(length(msg))\n\n$msg")
```
I'm  surprised using PackageCompiler makes such a big difference when the type of input isn't consistent between the compilation run and the normal run.

ffevotte replied:
Yes, I would say this is because most of the compilation/inference time is spent in sub-calls which work on parsed requests that always have the same type.

But this is only an intuition and I don't really know for sure. I'm just glad such an easy solution works so well, because I'm not sure how requests could be sent to the server via `stdin` while it is being package-compiled.

non-Jedi reviewed (approved):
First commit I'm good with. Still need to review second one.

non-Jedi commented on eglot-jl.el:
```diff
@@ -90,5 +116,20 @@ Otherwise returns nil"
 
+;;;###autoload
+(defun eglot-jl-compile-sysimage ()
+  "Create a compiled system image to reduce subsequent language
```
Please fit a single sentence description on the first line of docstrings.

ffevotte replied:
This should be done in the latest version.

non-Jedi commented on eglot-jl.el:
```diff
@@ -71,10 +76,31 @@ Otherwise returns nil"
+have been disabled altogether."
+  (when eglot-jl-enable-sysimage
+    (let ((julia-version (with-temp-buffer
+                           (call-process eglot-jl-julia-command nil t nil
```
Given the additional overhead of calling `julia`, I think my preference would be to default `eglot-jl-enable-sysimage` to nil. Obvious counter-argument is that there's no point in trying to avoid a call to julia when the language server itself takes 20+ seconds to start.

ffevotte replied:
About the overhead itself:

As it is written now, we're speaking of an overhead of a few hundreds of miliseconds (on my rather old system):
```
$ time julia -e 'print(VERSION)'
1.5.0-rc1.0

real    0m0,259s
user    0m0,328s
sys     0m0,203s
```

I had not thought about this until now, but this could be reduced by decreasing the compilation level:
```
$ time julia --compile=min -e 'print(VERSION)'
1.5.0-rc1.0

real    0m0,154s
user    0m0,077s
sys     0m0,070s
```

If we want to reduce this overhead further, the best option I can think of would be to use `julia --version`:
```
$ time julia --version
julia version 1.5.0-rc1

real    0m0,039s
user    0m0,020s
sys     0m0,001s
```
To be honest, the reason why I did not chose this option was because this involves additional parsing on the output but:
1. I was lazy and wanted to avoid doing it if possible, and
2. I did not know whether this format is normalized and consistently used across all julia versions.

One more word about this sub-issue regarding whether we should spawn a fully-fledged julia process or just use `julia --version`: the former is more flexible in that it allows doing more things in Julia, such as getting the versions of all packages, computing the correct file extension or other things like that. (More on that later in the answers to your other comments)

ffevotte replied:
On the broader discussion regarding the opt-in / opt-out mechanism:

Like you said, my initial rationale was that this "auto-detection" feature imposes an overhead that is completely negligible for users who don't opt in the feature, and acceptable to users who opt in.

I'd like to (if possible) avoid a situation where users who want to opt in would have to perform two manual actions: change a setting in their config file, and trigger the generation of the sysimage. Maybe an alternative mechanism would consist in having users opt-in by setting `eglot-jl-enable-sysimage` in their config file, and then `eglot-jl` would try and detect an existing, appropriate sysimage, use it if found or automatically compile one if none is found. What would you think of that?

non-Jedi commented on eglot-jl.el:
```diff
@@ -71,10 +76,31 @@ Otherwise returns nil"
+                           (buffer-string))))
+      (car
+       (remove nil
+               (mapcar
```
I think this code is much clearer if it simply checks `system-type` and then looks specifically for the appropriate file-extension rather than checking all of them.

ffevotte replied:
You're right. Among the options I thought of, my personal favorite would be to implement the logic of the sysimage name construction once, and re-use it both in `compile.jl` and `eglot-jl.el`. Unless I'm missing something, this would be feasible only if the "master implementation" of this logic is in Julia, which increases a bit the overhead of spawning a julia process in order to determine the sysimage name:
```
$ time julia --compile=min -e 'include("utils.jl"); print(sysimage_path())'                                                          
/home/francois/projets/emacs/eglot-jl/eglot-jl.1.5.0-rc1.0.so
real    0m0,256s
user    0m0,288s
sys     0m0,112s
```

However, if we want to reduce the overhead to a minimum, I could of course do as you propose, and re-implement the exact same logic in elisp. In which case we'd better wait to be sure about which information we want to encode in the sysimage name.

All this makes this issue rather intricately linked to other points that you raise, so that our final take on this will probably come naturally from the choices we make on other issues.

non-Jedi commented on eglot-jl.el:
```diff
@@ -71,10 +76,31 @@ Otherwise returns nil"
+               (mapcar
+                (lambda (ext)
+                  (let ((path (expand-file-name
+                               (concat "eglot-jl." julia-version ext)
```
This doesn't encode the version of LanguageServer and other packages into the filename. What happens when we release a new version of `eglot-jl` with an updated Manifest? I don't actually know how custom sysimages and project manifests interact. Apologies if this is covered somewhere, and I missed it.

ffevotte replied:
This is a very important issue, and probably a complex one, too. So we should take some time to discuss it.

A first point to consider is that all users installing and using `eglot-jl` from MELPA should have each version installed in a separate directory. For such users, there is therefore no risk of having an older sysimage (i.e. previously compiled for an older version of `eglot-jl` & its deps), mess up anything in a newer version of `eglot-jl`. What happens now for such users is that as soon as they update `eglot-jl`, no sysimage is found and they have to manually compile a new one (or wait for the 20s start-up time). If sysimage compilation was made automatic upon opt-in (as per my proposal above), then after each `eglot-jl` update a new sysimage compilation would be automatically spawned in he background (or the user would be prompted to before this happens).

For users who do not install `eglot-jl` via MELPA, but rather clone the repository somewhere and load the package from there, the situation is entirely different. Updating the git repository would be performed in place, and leave any existing sysimage there to be discovered and used even if it is out-of-date. What happens is that when the julia version gets updated, the sysimage becomes invalid. When the Manifest gets updated, any previously compiled sysimage remains valid, but out-of-date in the sense that it does not incorporate new developments in `LanguageServer` and `SymbolServer`.

For such users, we would need a mechanism to invalidate the sysimage when the Manifest changes. And the only option we have is to change the name so that no appropriate system image is found. I can think of a few ways to do it:
- get the versions of dependencies from `Pkg` (e.g. from `Pkg.status()`) and somehow include them in the sysimage file name. This makes it impossible to compute the sysimage name in elisp: it has to be done in julia, which adds to the aforementioned overhead;
- work directly from the Manifest file itself, for example computing a hash of its contents and incorporating that in the sysimage name. This makes it possible to compute the sysimage name in a consistent way both in elisp (using `secure-hash`) and in julia (using *e.g.* `SHA.sha1`)

non-Jedi replied:
I think using the hash of the manifest or some truncated version of it is probably the cleanest solution here. I think that the borg package manager installs packages directly from their git repo, so it's probably a workflow that should be supported.

non-Jedi reviewed (commented):
I still need to spend a bit more time looking through this commit, but I realized that I would never get to this if I waited until I had time to review the whole thing, so consider this a partial review of the second commit.

ffevotte commented:
As always, thanks for your insightful comments, that are thought-provoking.

To sum everything up, my current position would be along the lines of:
- computing a system image name that incorporates both the Julia version and a SHA1 hash of the Manifest
- implementing this only in Julia, and re-using the implementation by calling Julia from elisp (but I'm open to having implementations both in elisp and in julia to reduce the overhead of spawning a julia process if you feel that this is desirable)
- switching to an opt-in mechanism solely based on the `eglot-jl-enable-sysimage` cusomization variable, that is set to nil by default so that users who don't opt-in would have absolutely no overhead. And users who opt in wouldn't have to manually call `eglot-jl-compile-sysimage`: if no suitable sysimage is found when `eglot` starts, they would be prompted to launch the compilation of one in the background.

What would you think about that?

non-Jedi commented:
That sounds like a good path forward to me. :) I will note that if using the custom sysimage requires spawning multiple julia processes, it should definitely be opt-in rather than opt-out, but that's the plan anyway.

As an alternative, maybe `eglot-jl-enable-sysimage` could cause the sysimage to be compiled on `eglot-jl-init` so that compiling the sysimage doesn't run afoul of `eglot-connect-timeout`? Then invoking `eglot` either errors or shows a warning if the sysimage isn't available?

non-Jedi commented on eglot-jl.el:
```diff
@@ -60,6 +60,11 @@ The project should have LanguageServer and SymbolServer packages
   :type 'string)
 
+(defcustom eglot-jl-enable-sysimage t
+  "If t, use a system image if a suitable one is
```
```suggestion
  "When non-nil attempt to use a PackageCompiler sysimage.
```

Another one that needs single-line summary.

non-Jedi commented on eglot-jl.el:
```diff
@@ -90,5 +116,20 @@ Otherwise returns nil"
+  "Create a compiled system image for the language server.
+This reduces subsequent start-up times."
+  (interactive)
+  (let ((buffer (get-buffer-create "*eglot-jl sysimage compilation*")))
```
You can probably use the `compile` emacs function which should cut down some of this boiler-plate and make this asynchronous for free.

ffevotte replied:
One of the upsides of re-using `compile` (or the low-level `compilation-start`) would be to get correct treatment of ansi color sequences.

However, one big downside (IMO) is that `compile` & co are designed to be used interactively by humans. As such, they handle shell commands presented in the form of a string that has to be parsed. In order for us to use such an API, we'd have to first quote all arguments and join them together, which I find less safe (and reduces the gain in terms of boilerplate)

We can however easily put the sysimage compilation buffer in `view-mode` in order to benefit from standard key bindings traditionally associated to read-only buffers (e.g. <kbd>q</kbd> to bury the buffer).

non-Jedi commented:
Related: https://github.com/JuliaLang/julia/pull/35794

ffevotte committed:
Opt-in mechanism + more robust sysimage naming convention

* Opt-in mechanism:
  - `eglot-jl-enable-sysimage` is nil by default. No additional latency is
    introduced unless this setting is manually changed.
  - Setting `eglot-jl-enable-sysimage` automatically triggers sysimage
    compilation on `eglot-jl-init` when no suitable sysimage is found.
  - If `eglot-jl-enable-sysimage` is set and a suitable system image is found,
    it is automatically used.

* System image naming convention:
  - new naming convention: ${DIR}/eglot-jl_${VERSION}_${SLUG}.{EXT}
    + ${DIR}: path to the active project (should be the one defined by
        `eglot-jl-language-server-project`). Defaults to `eglot-jl-base` for
        Julia versions <1.4 where Pkg.project() is not defined.
    + ${VERSION}: Julia version string. This should allow avoiding problems with
        the sysimage becoming incorrect when Julia gets updated.
    + ${SLUG}: sha1 checksum of the contents of ${DIR}/Manifest.toml. This
        should allow avoiding problems with the sysimage becoming out-of-date
        when eglot-jl and its Julia dependencies get updated.
    + ${EXT}: suitable extension depending on the OS.
  - names following this convention are computed by a small julia function
    defined in `sysimage-path.jl`. When needed in Emacs, a Julia process is
    spawned for this (never happens unless `eglot-jl-enable-sysimage` is set).

ffevotte commented:
I'v finally gotten around to working on this again. The latest commit should implement the strategy discussed above:

* Opt-in mechanism:
  - `eglot-jl-enable-sysimage` is nil by default. No additional latency is    introduced unless this setting is manually changed.
  - Setting `eglot-jl-enable-sysimage` automatically triggers sysimage    compilation on `eglot-jl-init` when no suitable sysimage is found.
  - If `eglot-jl-enable-sysimage` is set and a suitable system image is found,    it is automatically used.

* System image naming convention:
  - new naming convention: ${DIR}/eglot-jl_${VERSION}_${SLUG}.{EXT}
    + ${DIR}: path to the active project (should be the one defined by        `eglot-jl-language-server-project`). Defaults to `eglot-jl-base` for        Julia versions <1.4 where Pkg.project() is not defined.
    + ${VERSION}: Julia version string. This should allow avoiding problems with        the sysimage becoming incorrect when Julia gets updated.
    + ${SLUG}: sha1 checksum of the contents of ${DIR}/Manifest.toml. This        should allow avoiding problems with the sysimage becoming out-of-date        when eglot-jl and its Julia dependencies get updated.
    + ${EXT}: suitable extension depending on the OS.
  - names following this convention are computed by a small julia function    defined in `sysimage-path.jl`. When needed in Emacs, a Julia process is    spawned for this (never happens unless `eglot-jl-enable-sysimage` is set).

@non-Jedi When you have some time to look at this (no rush!), I'd love to hear what you think.

ffevotte commented:
There is a small gotcha regarding the use of system images in conjunction with customized `eglot-jl-language-server-project`: although the Emacs side of the process knows which LanguageServer project should be activated, we currently don't forward that explicitly to Julia. And introspection capabilities have only been recently introduced in Pkg, so that for older Julia versions, there is no way (that I know of) of retrieving the currently active project in order to correctly encode the SHA1 of its Manifest in the sysimage name.

There are several ways of dealing with this that I can think of:
- leave things like this, and document the fact that `eglot-jl-enable-sysimage` and `eglot-jl-language-server-project` only work well together for Julia 1.4 and above
- use an environment variable to forward `eglot-jl-language-server-project` to the Julia process; this would probably be the easiest solution to implement, requiring minimal changes to the current code. But "hiding" this in an environment variable does not look too clean to me.
- pass `eglot-jl-language-server-project` as an additional (optional) command-line argument to `eglot-jl.jl`. But I fear this will make calling `eglot-jl.jl` more complex, for maybe not too much of a benefit.

I think my preference goes to this last option, but I'd love to hear your thoughts about it if you have some...

ffevotte commented:
Note to my (not so distant) future self: I also still have to update `README.md`

ffevotte committed:
Make `enable-sysimage` and `language-server-project` compatible

Make `eglot-jl-enable-sysimage` and `eglot-jl-language-server-project`
compatible with each other in all cases, for all Julia versions.

This works by propagating the value of `eglot-jl-language-server-project` from
the Emacs side to the Julia side via command-line arguments when necessary:
- in order to determine the correct system image name to look for, and
- in order to compile the system image itself.

ffevotte commented:
I went ahead and fixed the issues with `eglot-jl-language-server-project` and `eglot-jl-enable-sysimage` being used together in the latest commit. I'm happy with this solution, since it does not affect `eglot-jl.jl` in any way for users not opting in the sysimage feature.

ffevotte committed:
Update README.org

non-Jedi commented:
I haven't had an opportunity to use Julia for a while to test-drive this, but at this point if you've been using this without issue, I'm fine with merging it. Sorry for the extremely long review delay.

ffevotte committed:
Resolve environments before compiling sysimages

ffevotte commented:
This new version is rebased on the current master. It also incorporates (minor) modifications related to the need to resolve environments when switching Julia version.

I'm planning to test this in daily life for a few days in order to check that nothing went wrong with those last changes.

chriselrod commented:
This would still be neice to have, even with the improvements in 1.9.

linwaytin commented:
Same here. The startup time is still quite long for me. Hopefully this can be merged.
